source/distributed/fully_distributed_tria_util.cc

```diff
@@ -0,0 +1,523 @@
+            }
+
+          unsigned int temp = coarse_cell_id;
+          for (auto i : cell_indices)
+            {
+              temp = temp * GeometryInfo<dim>::max_children_per_cell + i;
+            }
+
+          return temp;
+        }
+      } // namespace
+
+
+      template <int dim, int spacedim>
+      ConstructionData<dim, spacedim>
+      create_construction_data_from_triangulation(
+        const dealii::Triangulation<dim, spacedim> &tria,
+        const Triangulation<dim, spacedim> &        tria_pft,
+        const unsigned int                          my_rank_in)
+      {
+        const MPI_Comm comm = tria_pft.get_communicator();
+
+        if (auto tria_pdt = dynamic_cast<
+              const parallel::distributed::Triangulation<dim, spacedim> *>(
+              &tria))
+          AssertThrow(comm == tria_pdt->get_communicator(),
+                      ExcMessage("MPI communicators do not match."));
+
+        unsigned int my_rank = my_rank_in;
+        AssertThrow(my_rank == numbers::invalid_unsigned_int ||
+                      my_rank < dealii::Utilities::MPI::n_mpi_processes(comm),
+                    ExcMessage(
+                      "Rank has to be smaller than available processes."));
+
+        if (auto tria_pdt = dynamic_cast<
+              const parallel::distributed::Triangulation<dim, spacedim> *>(
+              &tria))
+          {
+            if (my_rank == numbers::invalid_unsigned_int ||
+                my_rank == dealii::Utilities::MPI::this_mpi_process(comm))
+              my_rank = dealii::Utilities::MPI::this_mpi_process(comm);
+            else
+              AssertThrow(
+                false,
+                ExcMessage(
+                  "If parallel::distributed::Triangulation as source triangulation, my_rank has to equal global rank."));
+          }
+        else if (auto tria_serial =
+                   dynamic_cast<const dealii::Triangulation<dim, spacedim> *>(
+                     &tria))
+          {
+            if (my_rank == numbers::invalid_unsigned_int)
+              my_rank = dealii::Utilities::MPI::this_mpi_process(comm);
+          }
+        else
+          {
+            AssertThrow(false,
+                        ExcMessage(
+                          "This type of triangulation is not supported!"));
+          }
+
+        ConstructionData<dim, spacedim> construction_data;
+
+        auto &cells    = construction_data.coarse_cells;
+        auto &vertices = construction_data.coarse_cell_vertices;
+        auto &coarse_cell_index_to_coarse_cell_id =
+          construction_data.coarse_cell_index_to_coarse_cell_id;
+        auto &cell_infos = construction_data.cell_infos;
+
+        // helper function, which collects all vertices belonging to a cell
+        // (also taking into account periodicity)
+        auto add_vertices_of_cell_to_vertices_owned_by_loclly_owned_cells =
+          [](auto &cell, auto &vertices_owned_by_loclly_owned_cells) mutable {
```

Comment:
- I used to need the `mutable` her since the lambda used to have a side-effect.
- It is really annoying that in C++11 lambda and auto do not work together!